File: macos/Sources/RelicCore/SaveAudit.swift
import Foundation

/// 一份存档的完整审计结果：解析结果 + 逐件审计 + 展示所需的词条名与说明。
///
/// 这是「存档检查」页、报告导出与存档对比共用的数据结构；构造只走
/// `SaveAuditPipeline`，保证三处口径一致。
public struct AuditedSave: Sendable {
    /// 一件遗物 + 它的审计结论。
    public struct AuditedRelic: Identifiable, Sendable {
        public let relic: SaveRelic
        public let info: RelicInfo?
        public let result: RelicAuditResult

        public var id: Int { relic.index }
        public var isDeep: Bool { info?.deep == true }
        public var displayName: String { relicDisplayName(id: relic.itemID, info: info) }
        public var kindLabel: String { relicKindLabel(id: relic.itemID, info: info) }
        /// 颜色文案；遗物表里查不到这件遗物时为 nil。
        public var colorLabel: String? { info.map { relicColorLabel($0.color) } }

        /// 「非法 / 警告 / 合法」三态文案（页面与报告统一用这一份口径）。
        public var statusLabel: String {
            if result.status == .invalid { return "非法" }
            return result.warnings.isEmpty ? "合法" : "警告"
        }

        public init(relic: SaveRelic, info: RelicInfo?, result: RelicAuditResult) {
            self.relic = relic
            self.info = info
            self.result = result
        }
    }

    /// 一个角色槽位。
    public struct Character: Identifiable, Sendable {
        public let slot: Int
        public let name: String
        public let parseError: String?
        public let relics: [AuditedRelic]

        public var id: Int { slot }

        public var displayName: String { "槽位 \(slot + 1) · \(name)" }

        public var invalidCount: Int { relics.filter { $0.result.status == .invalid }.count }

        public var warningCount: Int {
            relics.filter { $0.result.status != .invalid && !$0.result.warnings.isEmpty }.count
        }

        public init(slot: Int, name: String, parseError: String?, relics: [AuditedRelic]) {
            self.slot = slot
            self.name = name
            self.parseError = parseError
            self.relics = relics
        }
    }

    public let fileName: String
    public let checksumOk: Bool
    /// effectId → 词条名（affixes.json ∪ extraAffixes）。
    public let affixNames: [Int: String]
    /// effectId → 词条说明（只收录 affixes.json 里非空的 explanation）。
    public let affixExplanations: [Int: String]
    public let characters: [Character]

    public init(
        fileName: String,
        checksumOk: Bool,
        affixNames: [Int: String],
        affixExplanations: [Int: String] = [:],
        characters: [Character]
    ) {
        self.fileName = fileName
        self.checksumOk = checksumOk
        self.affixNames = affixNames
        self.affixExplanations = affixExplanations
        self.characters = characters
    }

    public var relicCount: Int { characters.reduce(0) { $0 + $1.relics.count } }

    public var invalidCount: Int { characters.reduce(0) { $0 + $1.invalidCount } }

    public func affixName(_ id: Int) -> String {
        if let name = affixNames[id], !name.isEmpty { return name }
        return "未知词条 #\(id)"
    }

    /// 词条说明；没有收录说明时为 nil（页面据此不显示）。
    public func affixExplanation(_ id: Int) -> String? {
        guard let text = affixExplanations[id], !text.isEmpty else { return nil }
        return text
    }

    /// 「词条名（ID）」，报告与对比列表统一用这一份写法。
    public func affixLabel(_ id: Int) -> String {
        id == -1 ? "（空）" : "\(affixName(id))（\(id)）"
    }
}

/// 解析结果 → 审计结果的装配流程（纯逻辑，可在检查里直接调用）。
public enum SaveAuditPipeline {
    /// effectId → 词条说明（只收录非空的 explanation）。
    ///
    /// 复用 `context` 重载时由调用方建一次、传进去，避免每份存档重复遍历词条库。
    public static func explanations(from catalog: AffixCatalog) -> [Int: String] {
        var explanations: [Int: String] = [:]
        for affix in catalog.affixes where !affix.explanation.isEmpty {
            explanations[affix.effectID] = affix.explanation
        }
        return explanations
    }

    /// 用词条库与遗物表审计一份已解析的存档。
    public static func audit(
        _ parsed: SaveParseResult,
        catalog: AffixCatalog,
        relicData: RelicCatalog
    ) -> AuditedSave {
        let context = RelicAuditContext(catalog: catalog, relicData: relicData)
        return audit(parsed, context: context, affixExplanations: explanations(from: catalog))
    }

    /// 复用已经建好的审计上下文（对比两份存档时避免重复建索引）。
    ///
    /// `affixExplanations` 必填：漏传会让界面上的词条说明（ⓘ）静默消失，
    /// 没有说明可用时显式传 `[:]`，需要说明就用 `explanations(from:)` 建一份。
    public static func audit(
        _ parsed: SaveParseResult,
        context: RelicAuditContext,
        affixExplanations: [Int: String]
    ) -> AuditedSave {
        let auditor = RelicAuditor()
        let characters = parsed.characters.map { character -> AuditedSave.Character in
            var results = character.relics.map { auditor.audit($0, context: context) }
            auditor.applyUniqueDuplicates(&results, relics: character.relics)
            let relics = zip(character.relics, results).map { relic, result in
                AuditedSave.AuditedRelic(
                    relic: relic,
                    info: context.relicsByID[relic.itemID],
                    result: result
                )
            }
            return AuditedSave.Character(
                slot: character.slot,
                name: character.name,
                parseError: character.parseError,
                relics: relics
            )
        }
        return AuditedSave(
            fileName: parsed.fileName,
            checksumOk: parsed.checksumOk,
            affixNames: context.affixIndex.mapValues(\.name),
            affixExplanations: affixExplanations,
            characters: characters
        )
    }
}

File: android/gamedata/src/main/kotlin/com/nightreign/relicchecker/gamedata/save/SaveAudit.kt
package com.nightreign.relicchecker.gamedata.save

import com.nightreign.relicchecker.catalog.AffixCatalog
import com.nightreign.relicchecker.rules.foldedForSearch

// 解析结果 → 审计结果（移植 macOS 端 RelicCore/SaveAudit.swift）。
// 存档检查页、报告导出与存档对比共用 AuditedSave，构造只走 SaveAuditPipeline，保证三处口径一致。

/** 一件遗物 + 它的审计结论。 */
class AuditedRelic(
    val relic: SaveRelic,
    val info: RelicInfo?,
    val result: RelicAuditResult,
    /** 预先折叠好的搜索文本（遗物名、ID、种类、全部正负词条名与 ID），由审计管线在后台线程算好。 */
    val searchText: String = "",
) {
    val isDeep: Boolean get() = info?.deep == true
    val displayName: String get() = relicDisplayName(relic.itemId, info)
    val kindLabel: String get() = relicKindLabel(relic.itemId, info)

    /** 颜色文案；遗物表里查不到时写「颜色未知」（遗物卡、TXT、CSV 同一写法）。 */
    val colorText: String get() = info?.let { relicColorLabel(it.color) + "色" } ?: "颜色未知"

    val isInvalid: Boolean get() = result.status == RelicAuditStatus.INVALID

    /** 「非法 / 警告 / 合法」三态（页面与报告同一口径）。 */
    val statusLabel: String
        get() = when {
            isInvalid -> "非法"
            result.warnings.isEmpty() -> "合法"
            else -> "警告"
        }
}

/** 一个角色槽位。 */
class AuditedCharacter(
    val slot: Int,
    val name: String,
    val parseError: String?,
    val relics: List<AuditedRelic>,
) {
    /** 「槽位 N · 角色名」；读不出名字时写「未命名」。 */
    val displayName: String get() = "槽位 ${slot + 1} · ${name.ifEmpty { "未命名" }}"

    val invalidCount: Int = relics.count { it.isInvalid }
    val warningCount: Int = relics.count { !it.isInvalid && it.result.warnings.isNotEmpty() }
    val deepCount: Int = relics.count { it.isDeep }

    /** 空白 parseError 不算解析失败（与 Windows 端 parseErrorOf 一致）。 */
    val hasParseError: Boolean get() = !parseError.isNullOrBlank()
}

/** 一份存档的完整审计结果：解析结果 + 逐件审计 + 展示用的词条名与说明。 */
class AuditedSave(
    val fileName: String,
    val checksumOk: Boolean,
    /** effectId → 词条名（词条库 ∪ extraAffixes）。 */
    val affixNames: Map<Long, String>,
    /** effectId → 词条说明（只收录词条库里非空的 explanation）。 */
    val affixExplanations: Map<Long, String> = emptyMap(),
    val characters: List<AuditedCharacter>,
) {
    val relicCount: Int = characters.sumOf { it.relics.size }
    val invalidCount: Int = characters.sumOf { it.invalidCount }

    /** 「警告」件数；当前审计器不产出警告，恒为 0，报告据此省略该字段。 */
    val warningCount: Int = characters.sumOf { it.warningCount }

    fun affixName(id: Long): String = affixNames[id]?.takeIf { it.isNotEmpty() } ?: "未知词条 #$id"

    /** 词条说明；没有收录时为 null（页面据此不显示 ⓘ）。 */
    fun affixExplanation(id: Long): String? = affixExplanations[id]?.takeIf { it.isNotEmpty() }

    /** 「词条名（ID）」，报告与对比统一写法；空词条写「（空）」。 */
    fun affixLabel(id: Long): String {
        val normalized = normalizeEffectId(id)
        return if (normalized == -1L) "（空）" else "${affixName(normalized)}（$normalized）"
    }
}

object SaveAuditPipeline {
    /** effectId → 词条说明（只收录非空的 explanation）。 */
    fun explanations(catalog: AffixCatalog): Map<Long, String> =
        catalog.affixes.filter { it.explanation.isNotEmpty() }.associate { it.effectId.toLong() to it.explanation }

    /** 用词条库与遗物表审计一份已解析的存档（测试与一次性调用用；页面复用 [SaveAuditData]）。 */
    fun audit(parsed: SaveParseResult, catalog: AffixCatalog, relicData: RelicDataSet): AuditedSave =
        audit(parsed, RelicAuditContext(catalog, relicData), explanations(catalog))

    fun audit(parsed: SaveParseResult, data: SaveAuditData): AuditedSave =
        audit(parsed, data.context, data.explanations)

    /**
     * 复用已经建好的审计上下文（对比两份存档时避免重复建索引）。
     * [affixExplanations] 必填：漏传会让界面上的词条说明（ⓘ）静默消失。
     */
    fun audit(parsed: SaveParseResult, context: RelicAuditContext, affixExplanations: Map<Long, String>): AuditedSave {
        val auditor = RelicAuditor()
        val names = context.affixNames
        val characters = parsed.characters.map { character ->
            val results = auditor.applyUniqueDuplicates(
                character.relics.map { auditor.audit(it, context) },
                character.relics,
            )
            AuditedCharacter(
                slot = character.slot,
                name = character.name,
                parseError = character.parseError,
                relics = character.relics.mapIndexed { index, relic ->
                    val info = context.relicsById[relic.itemId]
                    AuditedRelic(relic, info, results[index], searchText(relic, info, names))
                },
            )
        }
        return AuditedSave(
            fileName = parsed.fileName,
            checksumOk = parsed.checksumOk,
            affixNames = names,
            affixExplanations = affixExplanations,
            characters = characters,
        )
    }

    /** 遗物卡搜索文本：名称、ID、种类、全部正负词条名与 ID（与桌面端一致），预先折叠。 */
    fun searchText(relic: SaveRelic, info: RelicInfo?, affixNames: Map<Long, String>): String {
        val parts = mutableListOf(relicDisplayName(relic.itemId, info), relic.itemId.toString(), relicKindLabel(relic.itemId, info))
        for (id in relic.effects + relic.curses) {
            if (id == -1L) continue
            parts += affixNames[id]?.takeIf { it.isNotEmpty() } ?: "未知词条 #$id"
            parts += id.toString()
        }
        return parts.joinToString(" ").foldedForSearch()
    }
}

/** 存档检查页的过滤（与桌面端「全部 / 仅非法 / 深夜遗物」一致）。 */
enum class SaveFilter(val title: String) {
    ALL("全部"),
    INVALID_ONLY("仅非法"),
    DEEP_ONLY("深夜遗物"),
    ;

    fun apply(relics: List<AuditedRelic>, query: String): List<AuditedRelic> {
        val needle = query.foldedForSearch()
        return relics.filter { relic ->
            val kept = when (this) {
                ALL -> true
                INVALID_ONLY -> relic.isInvalid
                DEEP_ONLY -> relic.isDeep
            }
            kept && (needle.isEmpty() || relic.searchText.contains(needle))
        }
    }

    /** 过滤后为空时的标题与说明（与桌面端同一套文案）。 */
    fun emptyTitle(): String = if (this == INVALID_ONLY) "🎉 未发现不合法遗物" else "没有符合条件的遗物"

    fun emptyDetail(character: AuditedCharacter): String = when {
        this == DEEP_ONLY -> "该角色没有持有深夜遗物。"
        character.relics.isEmpty() -> "该角色没有持有任何遗物。"
        else -> "当前过滤条件下没有可显示的遗物。"
    }
}
